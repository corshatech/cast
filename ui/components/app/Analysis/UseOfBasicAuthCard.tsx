import React from 'react';
import {
  DataGrid,
  GridColDef,
  GridRenderCellParams,
} from '@mui/x-data-grid';

import { Analysis } from '@/lib/findings';
import { AnalysisCard } from './core';
import { FormattedDate } from '@/components/atoms/FormattedDate';
import { CsvExportButton } from './core/CsvExportButton';

const columns: GridColDef[] = [
  {
    field: 'Timestamp',
    headerName: 'Timestamp',
    renderCell(params: GridRenderCellParams<string>) {
      return params.value && <FormattedDate when={params.value}/>
    },
    width: 200,
  },
  { field: 'Src IP', headerName: 'Src IP' },
  { field: 'URI', headerName: 'URI', width: 400 },
  { field: 'Dest IP', headerName: 'Dest IP' },
  { field: 'Dest Port', headerName: 'Dest Port' },
];

export const UseOfBasicAuthCard: React.FC<Analysis<'use-of-basic-auth'>> = ({
  findings,
  reportedAt,
  ...otherProps
}) => {
  const data = (findings ?? []).map(({
    data: {
      inRequest: {
        at,
        srcIp,
        URI,
        destIp,
        destPort,
      },
    },
  }) => ({
    id: `${at}${srcIp}${URI}${destPort}`,
    'Timestamp': at,
    'Src IP': srcIp,
    'URI': URI,
    'Dest IP': destIp,
    'Dest Port': destPort,
  }))
  return <AnalysisCard
    reportedAt={reportedAt}
    exportButton={<CsvExportButton
      stripID
      data={data}
      filename={`${reportedAt}-UseOfBasicAuth.csv`}
    />}
    {...otherProps}
  >
    <div style={{height: '400px', width: '100%'}}>
    <DataGrid
      rows={data}
      columns={columns}
      pageSize={10}
      rowsPerPageOptions={[10]}
    />
    </div>
  </AnalysisCard>
}
