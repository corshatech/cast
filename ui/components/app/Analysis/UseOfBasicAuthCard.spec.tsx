import { render } from '@testing-library/react';
import { UseOfBasicAuthCard } from './UseOfBasicAuthCard';

const reportedAt = new Date().toISOString();

describe('UseOfBasicAuthCard', () => {
  it('renders all passed props', () => {
    const { asFragment } = render(<UseOfBasicAuthCard
      id="use-of-basic-auth"
      title="Test title"
      description="Test description"
      reportedAt={reportedAt}
      severity="high"
      findings={[{
        type: 'use-of-basic-auth',
        name: 'Reused Authentication',
        detectedAt: reportedAt,
        severity: 'high',
        data: {
          inRequest: {
            at: reportedAt,
            destIp: '1.1.1.1',
            destPort: '5432',
            proto: 'tcp',
            srcIp: '2.2.2.2',
            srcPort: '1234',
            URI: 'https://example.com/',
          },
        },
      }]}
    />);
    expect(asFragment()).toMatchSnapshot();
  });
});
