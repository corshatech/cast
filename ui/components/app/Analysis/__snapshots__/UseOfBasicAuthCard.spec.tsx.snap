// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UseOfBasicAuthCard renders all passed props 1`] = `
<DocumentFragment>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-bhp9pd-MuiPaper-root-MuiCard-root"
  >
    <div
      class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
    >
      <div
        class="flex flex-nowrap"
      >
        <h2
          class="leading-6 text-h2 font-poppins font-bold grow"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-colorError MuiSvgIcon-fontSizeMedium css-1wxstni-MuiSvgIcon-root"
            data-testid="ErrorIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm1 15h-2v-2h2v2zm0-4h-2V7h2v6z"
            />
          </svg>
           Test title
        </h2>
        <button
          aria-label="download data as CSV"
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary MuiIconButton-sizeMedium css-1kuq5xv-MuiButtonBase-root-MuiIconButton-root"
          data-mui-internal-clone-element="true"
          tabindex="0"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="FileDownloadOutlinedIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M18 15v3H6v-3H4v3c0 1.1.9 2 2 2h12c1.1 0 2-.9 2-2v-3h-2zm-1-4-1.41-1.41L13 12.17V4h-2v8.17L8.41 9.59 7 11l5 5 5-5z"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
      <p
        class="leading-6 text-body1 font-nunito max-w-prose my-2"
      >
        Test description
      </p>
      <table
        class="font-light my-2 text-zinc-400 border-separate border-spacing-x-4"
      >
        <tbody>
          <tr>
            <td>
              Updated:
            </td>
            <td>
              2023-01-17T13:12:00.000Z
            </td>
          </tr>
          <tr>
            <td>
              Issue Severity:
            </td>
            <td>
              high
            </td>
          </tr>
        </tbody>
      </table>
      <div
        style="height: 400px; width: 100%;"
      >
        <div
          aria-colcount="5"
          aria-multiselectable="false"
          aria-rowcount="2"
          class="MuiDataGrid-root MuiDataGrid-root--densityStandard css-1e2bxag-MuiDataGrid-root"
          role="grid"
        >
          <div>
            <div />
          </div>
          <div
            class="MuiDataGrid-main css-204u17-MuiDataGrid-main"
          >
            <div
              class="MuiDataGrid-columnHeaders css-z8fhq1-MuiDataGrid-columnHeaders"
              style="min-height: 56px; max-height: 56px; line-height: 56px;"
            >
              <div
                class="MuiDataGrid-columnHeadersInner MuiDataGrid-columnHeadersInner--scrollable css-gl260s-MuiDataGrid-columnHeadersInner"
                role="rowgroup"
                style="transform: translate3d(0px, 0px, 0px);"
              >
                <div
                  aria-rowindex="1"
                  class="css-yrdy0g-MuiDataGrid-columnHeaderRow"
                  role="row"
                >
                  <div
                    aria-colindex="1"
                    aria-label="Timestamp"
                    aria-sort="none"
                    class="MuiDataGrid-columnHeader MuiDataGrid-columnHeader--sortable"
                    data-field="Timestamp"
                    role="columnheader"
                    style="height: 56px; width: 200px; min-width: 200px; max-width: 200px;"
                    tabindex="0"
                  >
                    <div
                      class="MuiDataGrid-columnHeaderDraggableContainer"
                      draggable="false"
                    >
                      <div
                        class="MuiDataGrid-columnHeaderTitleContainer"
                      >
                        <div
                          class="MuiDataGrid-columnHeaderTitleContainerContent"
                        >
                          <div
                            aria-label=""
                            class="MuiDataGrid-columnHeaderTitle css-1jbbcbn-MuiDataGrid-columnHeaderTitle"
                            data-mui-internal-clone-element="true"
                          >
                            Timestamp
                          </div>
                        </div>
                        <div
                          class="MuiDataGrid-iconButtonContainer css-ltf0zy-MuiDataGrid-iconButtonContainer"
                        >
                          <button
                            aria-label="Sort"
                            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-1pe4mpk-MuiButtonBase-root-MuiIconButton-root"
                            tabindex="-1"
                            title="Sort"
                            type="button"
                          >
                            <svg
                              aria-hidden="true"
                              class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall MuiDataGrid-sortIcon css-ptiqhd-MuiSvgIcon-root"
                              data-testid="ArrowUpwardIcon"
                              focusable="false"
                              viewBox="0 0 24 24"
                            >
                              <path
                                d="M4 12l1.41 1.41L11 7.83V20h2V7.83l5.58 5.59L20 12l-8-8-8 8z"
                              />
                            </svg>
                            <span
                              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                            />
                          </button>
                        </div>
                      </div>
                      <div
                        class="MuiDataGrid-menuIcon"
                      >
                        <button
                          aria-controls=":r3:"
                          aria-haspopup="true"
                          aria-label="Menu"
                          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall MuiDataGrid-menuIconButton css-1pe4mpk-MuiButtonBase-root-MuiIconButton-root"
                          id=":r4:"
                          tabindex="-1"
                          title="Menu"
                          type="button"
                        >
                          <svg
                            aria-hidden="true"
                            class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
                            data-testid="TripleDotsVerticalIcon"
                            focusable="false"
                            viewBox="0 0 24 24"
                          >
                            <path
                              d="M12 8c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2zm0 2c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm0 6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z"
                            />
                          </svg>
                          <span
                            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                          />
                        </button>
                      </div>
                    </div>
                    <div
                      class="MuiDataGrid-columnSeparator MuiDataGrid-columnSeparator--sideRight"
                      style="min-height: 56px; opacity: 1;"
                    >
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiDataGrid-iconSeparator css-i4bv87-MuiSvgIcon-root"
                        data-testid="SeparatorIcon"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M11 19V5h2v14z"
                        />
                      </svg>
                    </div>
                  </div>
                  <div
                    aria-colindex="2"
                    aria-label="Src IP"
                    aria-sort="none"
                    class="MuiDataGrid-columnHeader MuiDataGrid-columnHeader--sortable"
                    data-field="Src IP"
                    role="columnheader"
                    style="height: 56px; width: 100px; min-width: 100px; max-width: 100px;"
                    tabindex="-1"
                  >
                    <div
                      class="MuiDataGrid-columnHeaderDraggableContainer"
                      draggable="false"
                    >
                      <div
                        class="MuiDataGrid-columnHeaderTitleContainer"
                      >
                        <div
                          class="MuiDataGrid-columnHeaderTitleContainerContent"
                        >
                          <div
                            aria-label=""
                            class="MuiDataGrid-columnHeaderTitle css-1jbbcbn-MuiDataGrid-columnHeaderTitle"
                            data-mui-internal-clone-element="true"
                          >
                            Src IP
                          </div>
                        </div>
                        <div
                          class="MuiDataGrid-iconButtonContainer css-ltf0zy-MuiDataGrid-iconButtonContainer"
                        >
                          <button
                            aria-label="Sort"
                            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-1pe4mpk-MuiButtonBase-root-MuiIconButton-root"
                            tabindex="-1"
                            title="Sort"
                            type="button"
                          >
                            <svg
                              aria-hidden="true"
                              class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall MuiDataGrid-sortIcon css-ptiqhd-MuiSvgIcon-root"
                              data-testid="ArrowUpwardIcon"
                              focusable="false"
                              viewBox="0 0 24 24"
                            >
                              <path
                                d="M4 12l1.41 1.41L11 7.83V20h2V7.83l5.58 5.59L20 12l-8-8-8 8z"
                              />
                            </svg>
                            <span
                              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                            />
                          </button>
                        </div>
                      </div>
                      <div
                        class="MuiDataGrid-menuIcon"
                      >
                        <button
                          aria-controls=":r6:"
                          aria-haspopup="true"
                          aria-label="Menu"
                          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall MuiDataGrid-menuIconButton css-1pe4mpk-MuiButtonBase-root-MuiIconButton-root"
                          id=":r7:"
                          tabindex="-1"
                          title="Menu"
                          type="button"
                        >
                          <svg
                            aria-hidden="true"
                            class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
                            data-testid="TripleDotsVerticalIcon"
                            focusable="false"
                            viewBox="0 0 24 24"
                          >
                            <path
                              d="M12 8c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2zm0 2c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm0 6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z"
                            />
                          </svg>
                          <span
                            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                          />
                        </button>
                      </div>
                    </div>
                    <div
                      class="MuiDataGrid-columnSeparator MuiDataGrid-columnSeparator--sideRight"
                      style="min-height: 56px; opacity: 1;"
                    >
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiDataGrid-iconSeparator css-i4bv87-MuiSvgIcon-root"
                        data-testid="SeparatorIcon"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M11 19V5h2v14z"
                        />
                      </svg>
                    </div>
                  </div>
                  <div
                    aria-colindex="3"
                    aria-label="URI"
                    aria-sort="none"
                    class="MuiDataGrid-columnHeader MuiDataGrid-columnHeader--sortable"
                    data-field="URI"
                    role="columnheader"
                    style="height: 56px; width: 400px; min-width: 400px; max-width: 400px;"
                    tabindex="-1"
                  >
                    <div
                      class="MuiDataGrid-columnHeaderDraggableContainer"
                      draggable="false"
                    >
                      <div
                        class="MuiDataGrid-columnHeaderTitleContainer"
                      >
                        <div
                          class="MuiDataGrid-columnHeaderTitleContainerContent"
                        >
                          <div
                            aria-label=""
                            class="MuiDataGrid-columnHeaderTitle css-1jbbcbn-MuiDataGrid-columnHeaderTitle"
                            data-mui-internal-clone-element="true"
                          >
                            URI
                          </div>
                        </div>
                        <div
                          class="MuiDataGrid-iconButtonContainer css-ltf0zy-MuiDataGrid-iconButtonContainer"
                        >
                          <button
                            aria-label="Sort"
                            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-1pe4mpk-MuiButtonBase-root-MuiIconButton-root"
                            tabindex="-1"
                            title="Sort"
                            type="button"
                          >
                            <svg
                              aria-hidden="true"
                              class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall MuiDataGrid-sortIcon css-ptiqhd-MuiSvgIcon-root"
                              data-testid="ArrowUpwardIcon"
                              focusable="false"
                              viewBox="0 0 24 24"
                            >
                              <path
                                d="M4 12l1.41 1.41L11 7.83V20h2V7.83l5.58 5.59L20 12l-8-8-8 8z"
                              />
                            </svg>
                            <span
                              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                            />
                          </button>
                        </div>
                      </div>
                      <div
                        class="MuiDataGrid-menuIcon"
                      >
                        <button
                          aria-controls=":r9:"
                          aria-haspopup="true"
                          aria-label="Menu"
                          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall MuiDataGrid-menuIconButton css-1pe4mpk-MuiButtonBase-root-MuiIconButton-root"
                          id=":ra:"
                          tabindex="-1"
                          title="Menu"
                          type="button"
                        >
                          <svg
                            aria-hidden="true"
                            class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
                            data-testid="TripleDotsVerticalIcon"
                            focusable="false"
                            viewBox="0 0 24 24"
                          >
                            <path
                              d="M12 8c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2zm0 2c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm0 6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z"
                            />
                          </svg>
                          <span
                            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                          />
                        </button>
                      </div>
                    </div>
                    <div
                      class="MuiDataGrid-columnSeparator MuiDataGrid-columnSeparator--sideRight"
                      style="min-height: 56px; opacity: 1;"
                    >
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiDataGrid-iconSeparator css-i4bv87-MuiSvgIcon-root"
                        data-testid="SeparatorIcon"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M11 19V5h2v14z"
                        />
                      </svg>
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              style="overflow: visible; height: 0px; width: 0px;"
            >
              <div
                class="MuiDataGrid-virtualScroller css-1w5m2wr-MuiDataGrid-virtualScroller"
                style="width: 0px; height: auto; margin-top: 56px; overflow-x: hidden;"
              >
                <div
                  class="MuiDataGrid-virtualScrollerContent MuiDataGrid-virtualScrollerContent--overflowed css-1kwdphh-MuiDataGrid-virtualScrollerContent"
                  style="width: auto; height: 52px; min-height: auto;"
                >
                  <div
                    class="MuiDataGrid-virtualScrollerRenderZone css-s1v7zr-MuiDataGrid-virtualScrollerRenderZone"
                    style="transform: translate3d(0px, 0px, 0px);"
                  >
                    <div
                      aria-rowindex="2"
                      aria-selected="false"
                      class="MuiDataGrid-row MuiDataGrid-row--lastVisible"
                      data-id="2023-01-17T13:12:00.000Z2.2.2.2https://example.com/5432"
                      data-rowindex="0"
                      role="row"
                      style="max-height: 52px; min-height: 52px;"
                    >
                      <div
                        aria-colindex="1"
                        aria-colspan="1"
                        class="MuiDataGrid-cell--withRenderer MuiDataGrid-cell MuiDataGrid-cell--textLeft"
                        data-colindex="0"
                        data-field="Timestamp"
                        role="cell"
                        style="min-width: 200px; max-width: 200px; min-height: 52px; max-height: 52px;"
                        tabindex="-1"
                      >
                        2023-01-17T13:12:00.000Z
                      </div>
                      <div
                        aria-colindex="2"
                        aria-colspan="1"
                        class="MuiDataGrid-cell MuiDataGrid-cell--textLeft"
                        data-colindex="1"
                        data-field="Src IP"
                        role="cell"
                        style="min-width: 100px; max-width: 100px; min-height: 52px; max-height: 52px;"
                        tabindex="-1"
                      >
                        <div
                          class="MuiDataGrid-cellContent"
                          title="2.2.2.2"
                        >
                          2.2.2.2
                        </div>
                      </div>
                      <div
                        aria-colindex="3"
                        aria-colspan="1"
                        class="MuiDataGrid-cell MuiDataGrid-cell--textLeft"
                        data-colindex="2"
                        data-field="URI"
                        role="cell"
                        style="min-width: 400px; max-width: 400px; min-height: 52px; max-height: 52px;"
                        tabindex="-1"
                      >
                        <div
                          class="MuiDataGrid-cellContent"
                          title="https://example.com/"
                        >
                          https://example.com/
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="Mui-resizeTriggers"
            >
              <div
                class="expand-trigger"
              >
                <div
                  style="width: 1px; height: 1px;"
                />
              </div>
              <div
                class="contract-trigger"
              />
            </div>
          </div>
          <div>
            <div
              class="MuiDataGrid-footerContainer css-17jjc08-MuiDataGrid-footerContainer"
            >
              <div />
              <div
                class="MuiTablePagination-root css-rtrcn9-MuiTablePagination-root"
              >
                <div
                  class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular MuiTablePagination-toolbar css-78c6dr-MuiToolbar-root-MuiTablePagination-toolbar"
                >
                  <div
                    class="MuiTablePagination-spacer css-1psng7p-MuiTablePagination-spacer"
                  />
                  <p
                    class="MuiTablePagination-displayedRows css-levciy-MuiTablePagination-displayedRows"
                  >
                    1–1 of 1
                  </p>
                  <div
                    class="MuiTablePagination-actions"
                  >
                    <button
                      aria-label="Go to previous page"
                      class="MuiButtonBase-root Mui-disabled MuiIconButton-root Mui-disabled MuiIconButton-colorInherit MuiIconButton-sizeMedium css-zylse7-MuiButtonBase-root-MuiIconButton-root"
                      disabled=""
                      tabindex="-1"
                      title="Go to previous page"
                      type="button"
                    >
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                        data-testid="KeyboardArrowLeftIcon"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M15.41 16.09l-4.58-4.59 4.58-4.59L14 5.5l-6 6 6 6z"
                        />
                      </svg>
                    </button>
                    <button
                      aria-label="Go to next page"
                      class="MuiButtonBase-root Mui-disabled MuiIconButton-root Mui-disabled MuiIconButton-colorInherit MuiIconButton-sizeMedium css-zylse7-MuiButtonBase-root-MuiIconButton-root"
                      disabled=""
                      tabindex="-1"
                      title="Go to next page"
                      type="button"
                    >
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                        data-testid="KeyboardArrowRightIcon"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M8.59 16.34l4.58-4.59-4.58-4.59L10 5.75l6 6-6 6z"
                        />
                      </svg>
                    </button>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
